### PR TITLE
add spectral normalization [pytorch]

### PR DESCRIPTION
**Pull request by crcrpar**
related to #5027.
This PR aims at implementing Spectral Normalization in a way similar to `torch.nn.utils.weight_norm`.

### CODE REVIEW

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+from torch.nn.parameter import Parameter
+
+
+def l2normalize(v, eps=1e-12):
+    """Scale to inputs norm to 1."""
+    denom = v.norm(p=2) + eps
+    return v / denom
+
+
+class SpectralNorm(object):
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: This normalizes weights of layers, not outputs of layers.
So, I think it's preferable to implement like `WeightNorm` for flexibility.

File: torch/nn/utils/spectral_norm.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+
+def spectral_norm(module, name='weight', n_power_iterations=1, eps=1e-12):
+    r"""Applies spectral normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilize the training of discriminators(critics)
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: stabilizes, space before `(`

File: torch/nn/utils/spectral_norm.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    in GANs by rescaling the weight tensor by spectral norm "sigma" of the
+    weight matrix calculated by power iteration method. If the dimension of the
+    weight tensor is greater than 2, reahaped to 2D in power iteration method
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit reshaped

Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: thank you

File: torch/nn/utils/spectral_norm.py (hunk 2 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: thanks

File: torch/nn/utils/spectral_norm.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: @ezyang What do you think about this?

File: torch/nn/utils/spectral_norm.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
+"""
+Spectral Normalization from https://arxiv.org/abs/1802.05957
+"""
+import torch
+from torch.nn.parameter import Parameter
+
+
+def l2normalize(v, eps=1e-12):
```
Comment (fmassa) [inline torch/nn/utils/spectral_norm.py]: Can't you use instead [`torch.nn.functional.normalize`](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L2004)?

File: torch/nn/utils/spectral_norm.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
+        self.name = name
+        self.n_power_iterations = n_power_iterations
+        self.eps = eps
+
+    def compute_weight(self, module):
+        weight = module._parameters[self.name + '_org']
+        u = module._buffers[self.name + '_u']
+        height, _cuda = weight.size(0), weight.is_cuda
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: This is my fault and I'll remove it.

File: torch/nn/utils/spectral_norm.py (hunk 4 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: I mimic the author's [implementation](https://github.com/pfnet-research/sngan_projection/blob/master/source/functions/max_sv.py#L5) where the denominator is `norm + eps`, not `max(norm, eps)`.

I'm not sure which is suitable so if you think `F.normalize` better, I'll replace.

Comment (fmassa) [reply torch/nn/utils/spectral_norm.py]: I think there is no real difference actually. This is there just to avoid division by zero.

File: torch/nn/utils/spectral_norm.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+def spectral_norm(module, name='weight', n_power_iterations=1, eps=1e-12):
+    r"""Applies spectral normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor by spectral norm "sigma" of the
+    weight matrix calculated by power iteration method. If the dimension of the
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: this sentence has three `by`s chained together. How about 
```
Spectral normalization stabilizes the training of discriminators (critics)
in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of the 
weight matrix calculated using power iteration method.
```

File: torch/nn/utils/spectral_norm.py [hunk 7]
```diff
@@ -21,11 +16,11 @@ def __init__(self, name='weight', n_power_iterations=1, eps=1e-12):
     def compute_weight(self, module):
         weight = module._parameters[self.name + '_org']
         u = module._buffers[self.name + '_u']
-        height, _cuda = weight.size(0), weight.is_cuda
+        height = weight.size(0)
         weight_mat = weight.view(height, -1)
         for _ in range(self.n_power_iterations):
-            v = l2normalize(torch.matmul(weight_mat.t(), u), self.eps)
-            u = l2normalize(torch.matmul(weight_mat, v), self.eps)
+            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: Yes, `torch.matmul(weight_mat.t(), u)` returns 1d.
Since the arguments of `torch.nn.functional.normalize` are `input, p=2, dim=1, eps=1e-12`, I explicitly set `dim=0`.
Or, should I use `F.normalize`?

[the remark this answers is not in the document]
Comment (fmassa) [reply torch/nn/utils/spectral_norm.py]: using `torch.nn.functional.normalize` is fine then. I just wanted to make sure that we were not changing the behavior. Thanks!

**Comment by ezyang** [comment]
@fmassa @SsnL is this ok to merge now?

File: torch/nn/utils/spectral_norm.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    r"""Applies spectral normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of
+    the weight matrix calculated using power iteration method. If the dimension
+    of the weight tensor is greater than 2, reshaped to 2D in power iteration method
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: `it is reshaped`

File: torch/nn/utils/spectral_norm.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    to get spectral norm. This is implemented via a hook that calculates
+    spectral norm and rescales weight before every :meth:`~Module.forward`
+    call.
+
+    See https://arxiv.org/abs/1802.05957
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: I know that this is how weight_norm doc is. But could you write this as:
```
See `Spectral Normalization for Generative Adversarial Networks`_ .

.. _`Spectral Normalization for Generative Adversarial Networks`: https://arxiv.org/abs/1802.05957
```
? We should update that as well.

File: torch/nn/utils/spectral_norm.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    SpectralNorm.apply(module, name, n_power_iterations, eps)
+    return module
+
+
+def remove_spectral_norm(module, name='weight'):
+    r"""Removes the spectral normalization reparameterization from a module.
+
+    Args:
+        module (nn.Modue): containing module
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: `Module`

File: torch/nn/utils/spectral_norm.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        setattr(module, self.name, weight)
+        setattr(module, self.name + '_u', u)
+
+    @staticmethod
+    def apply(module, name, n_power_iterations, eps):
+        fn = SpectralNorm(name, n_power_iterations, eps)
+        weight = module._parameters[name]
+        height = weight.size(0)
+
+        u = normalize(weight.data.new(height).normal_(0, 1), dim=0, eps=fn.eps)
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: `.data` is not needed, and please use `new_empty` instead of `new`.

File: torch/nn/utils/spectral_norm.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+def spectral_norm(module, name='weight', n_power_iterations=1, eps=1e-12):
+    r"""Applies spectral normalization to a parameter in the given module.
+
+    .. math::
+         \mathbf{W} &= \dfrac{\mathbf{W}}{\sigma(\mathbf{W})} \\
+         \sigma(\mathbf{W}) &= \max_{\mathbf{h}: \mathbf{h} \ne 0} \dfrac{\|\mathbf{W} \mathbf{h}\|_2}{\|\mathbf{h}\|_2}
+
+    Spectral normalization stabilizes the training of discriminators (critics)
+    in GANs by rescaling the weight tensor with spectral norm :math:`\sigma` of
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: I'd suggest writing it out as `Generative Adversarial Networks (GANs)`.

File: torch/nn/utils/spectral_norm.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        self.name = name
+        self.n_power_iterations = n_power_iterations
+        self.eps = eps
+
+    def compute_weight(self, module):
+        weight = module._parameters[self.name + '_org']
+        u = module._buffers[self.name + '_u']
+        height = weight.size(0)
+        weight_mat = weight.view(height, -1)
+        for _ in range(self.n_power_iterations):
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: Can you add a comment like:
```
Spectral norm of weight equals to `u^T W v`, where `u` and `v` are the first left and right singular vectors.
This power iteration produces approximations of `u` and `v`.
```

File: torch/nn/utils/spectral_norm.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    call.
+
+    See https://arxiv.org/abs/1802.05957
+
+    Args:
+        module (nn.Module): containing module
+        name (str, optional): name of weight parameter
+        n_power_iterations (int, optional): number of power iterations to
+            calculate spectal norm
+        eps (float, optional): epsilon for numerical stability
```
Comment (SsnL) [inline torch/nn/utils/spectral_norm.py]: nit: please be specific, like `... for numerical stability in calculating norms`.

**Review by ssnl — COMMENTED** [review]
Code looks solid. There are still some minor nits around doc and comments. I think this is good to go once those are fixed.

File: torch/nn/utils/spectral_norm.py (hunk 8 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: thanks

File: torch/nn/utils/spectral_norm.py (hunk 11 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: I didn't know that method, thank you.

File: torch/nn/utils/spectral_norm.py (hunk 9 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: sure

File: torch/nn/utils/spectral_norm.py (hunk 12 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: OK, I'll update.

File: torch/nn/utils/spectral_norm.py (hunk 14 above)
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: I got it.

**Review by ssnl — COMMENTED** [review]
LGTM except the typo

File: torch/nn/utils/spectral_norm.py [hunk 15]
```diff
@@ -19,6 +19,9 @@ def compute_weight(self, module):
         height = weight.size(0)
         weight_mat = weight.view(height, -1)
         for _ in range(self.n_power_iterations):
+            # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+            # are the first left and right singular vecors.
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply torch/nn/utils/spectral_norm.py]: I'm sorry.

**Review by ssnl — APPROVED** [review]
LGTM. @fmassa , do you want to take an extra look?

**Review by ezyang — APPROVED** [review]

**Comment by SsnL** [comment]
Thank you! @crcrpar

**Comment by crcrpar** [comment]
Thank you for your very kind reviews and suggestions!

File: torch/nn/utils/spectral_norm.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        weight_mat = weight.view(height, -1)
+        for _ in range(self.n_power_iterations):
+            # Spectral norm of weight equals to `u^T W v`, where `u` and `v`
+            # are the first left and right singular vectors.
+            # This power iteration produces approximations of `u` and `v`.
+            v = normalize(torch.matmul(weight_mat.t(), u), dim=0, eps=self.eps)
+            u = normalize(torch.matmul(weight_mat, v), dim=0, eps=self.eps)
+
+        sigma = torch.dot(u, torch.matmul(weight_mat, v))
+        weight.data /= sigma
```
Comment (ferrine) [inline torch/nn/utils/spectral_norm.py]: Why is it made non differentiable?

Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: A couple bugs of this initial implementation is fixed before release, including making it differentials.
